askama_parser/src/lib.rs

```diff
@@ -350,13 +350,100 @@ fn str_lit(i: &str) -> ParseResult<'_> {
     Ok((i, s.unwrap_or_default()))
 }
 
+// Information about allowed character escapes is available at:
+// <https://doc.rust-lang.org/reference/tokens.html#ascii-escapes>.
 fn char_lit(i: &str) -> ParseResult<'_> {
+    let start = i;
     let (i, s) = delimited(
         char('\''),
         opt(escaped(is_not("\\\'"), '\\', anychar)),
         char('\''),
     )(i)?;
-    Ok((i, s.unwrap_or_default()))
+    let Some(s) = s else {
+        return Err(nom::Err::Failure(ErrorContext {
+            input: start,
+            // Same error as rustc.
+            message: Some(Cow::Borrowed("empty character literal")),
+        }));
+    };
+    if s.chars().count() == 1 {
+        // Nothing to check, all good.
+        return Ok((i, s));
+    }
+    let res: ParseResult<'_> = recognize(tuple((
+        char('\\'),
+        alt((
+            recognize(char('n')),
+            recognize(char('r')),
+            recognize(char('t')),
+            recognize(char('\\')),
+            recognize(char('0')),
+            recognize(char('\'')),
+            recognize(char('"')), // Not useful but supported by rust.
+            recognize(tuple((
+                char('x'),
+                take_while_m_n(2, 2, |c: char| c.is_ascii_hexdigit()),
+            ))),
+            recognize(tuple((
+                tag("u{"),
+                take_while_m_n(1, 6, |c: char| c.is_ascii_hexdigit()),
+                char('}'),
+            ))),
+        )),
+    )))(s);
+    let validate_endpoint = |prefix: &str,
+                             strip_last_bracket: bool,
+                             max_value: u32,
+                             err1: &'static str,
+                             err2: &'static str|
+     -> Result<(), nom::Err<ErrorContext<'_>>> {
+        let Some(mut nb) = s.strip_prefix(prefix) else {
+            return Ok(());
+        };
+        if strip_last_bracket {
+            if let Some(without_bracket) = nb.strip_suffix('}') {
+                nb = without_bracket;
+            }
+        }
+        let Ok(nb) = u32::from_str_radix(nb, 16) else {
+            return Err(nom::Err::Failure(ErrorContext {
+                input: start,
+                message: Some(Cow::Borrowed(err1)),
+            }));
+        };
+        if nb > max_value {
+            return Err(nom::Err::Failure(ErrorContext {
+                input: start,
+                message: Some(Cow::Borrowed(err2)),
+            }));
+        }
+        Ok(())
+    };
```

Comment:
The only variable this is capturing is `s`, right? I'd prefer to make this a real function with a docstring to explain what it tries to do and what the arguments/return value mean.

Reply:
`start` as well. I didn't make it a function because it's a lot of arguments but I can if you prefer.

Reply:
Also I think you don't need to take `strip_last_bracket`, can check whether the prefix ends with `{`.

Reply:
It's a (small) performance impact that can be avoided, so I preferred to rely on a boolean instead. I can also do it as such if you prefer though.

Reply:
Yes, I'd prefer to avoid it. I think we're into edge cases of edge cases here, so making minor optimizations like this definitely feels like premature optimization.

Reply:
:+1: